src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -36,19 +38,107 @@ internal static bool IsEnabled()
             //from DiagnosticListener right after the check. So some requests happening right after subscription starts
             //might not be instrumented. Similarly, when consumer unsubscribes, extra requests might be instumented
 
-            Guid loggingRequestId = Guid.NewGuid();
-            LogHttpRequest(request,  loggingRequestId);
+            Task<HttpResponseMessage> responseTask;
 
-            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
-
-            LogHttpResponse(responseTask, loggingRequestId);
+            //We need to know if activity events are enabled and send Request/Response events otherwise
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityName))
+            {
+                if (Activity.Current != null)
+                {
+                    responseTask = SendInstrumentedAsync(request, cancellationToken);
+                }
+                else
+                {
+                    //null Activity.Current means that incoming request was not instrumented
+                    //and there is nothing we can do for outgoing request. 
+                    //Activity events are enabled so we don't send Request/Response events
+                    responseTask = base.SendAsync(request, cancellationToken);
+                }
+            }
+            else 
+            {
+                Guid loggingRequestId = Guid.NewGuid();
+                LogHttpRequest(request, loggingRequestId);
+                responseTask = base.SendAsync(request, cancellationToken);
+                LogHttpResponse(responseTask, loggingRequestId);
+            }
             return responseTask;
         }
 
         #region private
 
-        private static readonly DiagnosticListener s_diagnosticListener =
-            new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
+        private async Task<HttpResponseMessage> SendInstrumentedAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Activity activity = null;
+            //check if user wants THIS request to be instrumented
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityName, request))
+            {
+                activity = new Activity(DiagnosticsHandlerLoggingStrings.ActivityName);
+
+                s_diagnosticListener.StartActivity(activity, new { Request = request });
+
+                //inject headers unless user does not want it to happen
+                //we may pass request to IsEnabled, but not clear how useful it is
+                if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityPropagateName))
```

Comment:
My bias is to NOT have the complexity of a Activity.Propagate check.   My concern is that it is just one more thing that loggers have to do for the system to work as expected (and it is a bit less efficient, and the code is a bit more complex).  

The rationale here is that putting users have control over propagation when things come into a machine, do they really ALSO need control when it leaves?

My preference is to remove this and if we find the need for this control we can add a 'NoPropagate' check at that time (my guess is we never will) and the system will end up simpler.     Even if we do end up adding it,  least that way the default will be good (by default you get propagation and you have to go out of your way to do something surprising).